drivers/oled/oled_driver.c

```diff
@@ -132,24 +168,116 @@ uint32_t oled_scroll_timeout;
 uint16_t oled_update_timeout;
 #endif
 
-// Internal variables to reduce math instructions
+#if defined(OLED_TRANSPORT_SPI)
+#    ifndef OLED_DC_PIN
+#        error "The OLED driver in SPI needs a D/C pin defined"
+#    endif
+#    ifndef OLED_CS_PIN
+#        error "The OLED driver in SPI needs a CS pin defined"
+#    endif
+#    ifndef OLED_SPI_MODE
+#        define OLED_SPI_MODE 3
+#    endif
+#    ifndef OLED_SPI_DIVISOR
+#        define OLED_SPI_DIVISOR 2
+#    endif
+#elif defined(OLED_TRANSPORT_I2C)
+#    if !defined(OLED_DISPLAY_ADDRESS)
+#        define OLED_DISPLAY_ADDRESS 0x3C
+#    endif
+#endif
+
+// Transmit/Write Funcs.
+__attribute__((weak)) bool oled_cmd(const uint8_t *data, uint16_t size) {
+#if defined(OLED_TRANSPORT_SPI)
+    if (!spi_start(OLED_CS_PIN, false, OLED_SPI_MODE, OLED_SPI_DIVISOR)) {
+        return false;
+    }
+    // Command Mode
+    writePinLow(OLED_DC_PIN);
+    // Send the commands
+    if (spi_transmit(&data[1], size - 1) != SPI_STATUS_SUCCESS) {
+        spi_stop();
+        return false;
+    }
+    spi_stop();
+    return true;
+#elif defined(OLED_TRANSPORT_I2C)
+    i2c_status_t status = i2c_transmit((OLED_DISPLAY_ADDRESS << 1), data, size, OLED_I2C_TIMEOUT);
```

Comment:
One possible refactoring here would be to use `i2c_writeReg((OLED_DISPLAY_ADDRESS << 1), I2C_CMD, data, size, OLED_I2C_TIMEOUT)` (and an explicit `i2c_write(I2C_CMD)` in the `_P` version), so that the driver implementation won't have `I2C_CMD` in the command buffers, and the SPI transport functions won't need to ignore the first byte in the buffer.  However, this would be slightly less efficient on ARM (the ChibiOS implementation of `i2c_writeReg()` allocates a buffer on the stack to add the register address byte before the data), so we may just accept this difference between the command and data transfer functions.

Reply:
another option was also if-defs for all the commands.  This was pretty much the cleanest implementation.